File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        vector<int> tmp = {k, 0};

        kthSmallestHelper(root, tmp);

        return tmp[1];
    }

    void kthSmallestHelper(TreeNode* root, vector<int>& tmp) {
        if (root == nullptr) {
            return;
        }

        kthSmallestHelper(root->left, tmp);

        if (tmp[0] <= 0) return;

        tmp[0]--;
        if (tmp[0] == 0) {
            tmp[1] = root->val;
            return;
        }

        kthSmallestHelper(root->right, tmp);
    }
};

File: Data Structures & Algorithms/sum-root-to-leaf-numbers/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def sumNumbers(self, root: Optional[TreeNode]) -> int:
        total, currNum = 0, 0

        def dfs(node):
            nonlocal total
            nonlocal currNum

            if not node:
                return

            currNum = currNum * 10 + node.val
            if not node.left and not node.right:
                total += currNum
                currNum //= 10
                return
            
            dfs(node.left)
            dfs(node.right)
            currNum //= 10

        dfs(root)
        return total